### PR TITLE
Fixes #2384 aws_cloudwatch_log_metric_filter: default_value is automatically set to 0

### LINKED ISSUE #2384: aws_cloudwatch_log_metric_filter: default_value is automatically set to 0
When I apply a terraform configuration that does NOT include default_value, terraform creates a log filter in AWS that has the default_value set to 0.

### Terraform Version
Terraform v0.10.3 with aws provider v1.3.1

### Affected Resource(s)
- aws_cloudwatch_log_metric_filter

### Terraform Configuration Files
```hcl
resource "aws_cloudwatch_log_metric_filter" "my-function-memory-used" {
  name           = "MyFunction_MemoryUsed"
  pattern        = "[type=REPORT, ... , memory_used, unit]"
  log_group_name = "/aws/lambda/MyFunction"

  metric_transformation {
    name      = "MemoryUsed"
    namespace = "MyFunction.CustomMetrics"
    value     = "$memory_used"
  }
}
```

### Expected Behavior
Creates a log filter with default_value **not set**, so that log lines that don't match the filter don't output any value for the metric. http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/MonitoringLogData.html

### Actual Behavior
Creates a log filter with default_value **set to 0**, so that every log line outputs a value for the metric (non-matching lines output 0). 

### Steps to Reproduce
1. Write terraform for a log_filter WITHOUT a default_value
1. `terraform apply` (note that the console output does not mention default_value at all)
    ```
    aws_cloudwatch_log_metric_filter.my-function-memory-used: Creating...
    [11:12:27][Step 2/2]   log_group_name:                    "" => "/aws/lambda/MyFunction"
    [11:12:27][Step 2/2]   metric_transformation.#:           "" => "1"
    [11:12:27][Step 2/2]   metric_transformation.0.name:      "" => "MemoryUsed"
    [11:12:27][Step 2/2]   metric_transformation.0.namespace: "" => "MyFunction.CustomMetrics"
    [11:12:27][Step 2/2]   metric_transformation.0.value:     "" => "$memory_used"
    [11:12:27][Step 2/2]   name:                              "" => "MyFunction_MemoryUsed"
    [11:12:27][Step 2/2]   pattern:                           "" => "[type=REPORT, ... , memory_used, unit]"
    ```
1. Visit the AWS Console CloudWatch and see that your new metric has a minimum value of 0 and an unexpectedly low average. 

### References
- #1578 - this PR from a month ago added support for the optional default_value

### PR DESCRIPTION
**Pull request by ahl**
Fixes #2384

Note I looked at #2759 and #2902 which were both old and incomplete
cc: @jocelynthode @modax 

Changes proposed in this pull request:

* use sentinel value of NaN to indicate that no value was specified
* clean up misspelling and cruft

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchLogMetricFilter_*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSCloudWatchLogMetricFilter_* -timeout 120m
=== RUN   TestAccAWSCloudWatchLogMetricFilter_basic
--- PASS: TestAccAWSCloudWatchLogMetricFilter_basic (18.69s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	18.743s

```

### CODE REVIEW

**Comment by ahl** [comment]
I should also have noted that a better solution is blocked on this issue: https://github.com/hashicorp/terraform/issues/16796

**Comment by ahl** [comment]
maybe @bflad you could take a look?

**Comment by paultyng** [comment]
I think we should implement this instead as a `TypeString` so that empty string can represent the unset value. We have precedent for this with `TypeBool`'s https://github.com/terraform-providers/terraform-provider-aws/blob/502d7b4cfa74d5a0715e6ba32f60d4ad58e2cbd7/aws/resource_aws_launch_template.go#L92-L95

It's definitely an annoying limitation of the current type system, and potentially something we can address post 0.12, but the backwards compatible implementation will probably just use a `TypeString` as well.

**Comment by ahl** [comment]
Makes sense; I'll update the PR

**Comment by ahl** [comment]
@jocelynthode I didn't know that was possible; thanks.

File: aws/validators.go [hunk 1]
```diff
@@ -54,6 +54,22 @@ func validateTypeStringNullableBoolean(v interface{}, k string) (ws []string, es
...
+// validateTypeStringNullableFloat provides custom error messaging for TypeString floats
+// Some arguments require a floating point value or an unspecified, empty field.
+func validateTypeStringNullableFloat(v interface{}, k string) (ws []string, es []error) {
+	value, ok := v.(string)
+	if !ok {
+		es = append(es, fmt.Errorf("expected type of %s to be string", k))
+		return
+	}
+
+	if _, err := strconv.ParseFloat(value, 64); err != nil {
```
Comment (paultyng) [inline aws/validators.go]: Should there be a check for empty string in here? May help to do a little unit testing on this as there are some edges. This seems to error if empty string.

https://play.golang.org/p/iBa1Pq5V2tt

**Comment by paultyng** [comment]
@ahl thanks for following up on this one!

**Comment by ahl** [comment]
@paultyng it's pretty critical to us so I was motivated; thanks for sticking with me on it.

**Comment by bflad** [comment]
This has been released in [version 1.44.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1440-november-14-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**